spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/source/ExtractRowLineage.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.spark.source;
+
+import java.util.List;
+import java.util.function.Function;
+import org.apache.iceberg.MetadataColumns;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.catalyst.ProjectingInternalRow;
+import org.apache.spark.sql.catalyst.expressions.GenericInternalRow;
+import org.apache.spark.sql.types.LongType$;
+import org.apache.spark.sql.types.StructType;
+import scala.collection.JavaConverters;
+
+class ExtractRowLineage implements Function<InternalRow, InternalRow> {
+  private static final StructType ROW_LINEAGE_SCHEMA =
+      new StructType()
+          .add(MetadataColumns.ROW_ID.name(), LongType$.MODULE$, true)
+          .add(MetadataColumns.LAST_UPDATED_SEQUENCE_NUMBER.name(), LongType$.MODULE$, true);
+  private static final InternalRow EMPTY_LINEAGE_ROW = new GenericInternalRow(2);
+
+  private final Schema writeSchema;
+
+  private ProjectingInternalRow cachedRowLineageProjection;
+
+  ExtractRowLineage(Schema writeSchema) {
+    Preconditions.checkArgument(writeSchema != null, "Write schema cannot be null");
+    this.writeSchema = writeSchema;
+  }
+
+  @Override
+  public InternalRow apply(InternalRow meta) {
+    // If row lineage is not required on write return a null row
+    if (writeSchema.findField(MetadataColumns.ROW_ID.name()) == null) {
```

Comment:
since `writeSchema` can't be null. can we do the check in constructor and save the result in a boolean flag to avoid repeated `findField`?

since this is in a hot code path. I am a little extra careful / paranoid :)

Reply:
I updated to have a boolean `rowLineageRequiredField` which we only set once in the constructor and then it's just a boolean check on every record